containers-tests/tests/map-properties.hs

```diff
@@ -1222,31 +1228,68 @@ prop_mergeA_effects xs ys
 
 ----------------------------------------------------------------
 
-prop_ordered :: Property
-prop_ordered
-  = forAll (choose (5,100)) $ \n ->
-    let xs = [(x,()) | x <- [0..n::Int]]
-    in fromAscList xs == fromList xs
-
-prop_rev_ordered :: Property
-prop_rev_ordered
-  = forAll (choose (5,100)) $ \n ->
-    let xs = [(x,()) | x <- [0..n::Int]]
-    in fromDescList (reverse xs) == fromList xs
+-- fromAscListWith, fromAscListWithKey, fromDescListWith, fromDescListWithKey
+-- all effectively perform a left fold over adjacent elements in the input list
+-- using some function as long as the keys are equal.
+--
+-- The property tests for these functions compare the result against the
+-- sequence we would get if we used NE.groupBy instead. We use Magma to check
+-- the fold direction (left, not right) and the order of arguments to the fold
+-- function (new then old).
+
+data Magma a
+  = Inj a
+  | Magma a :* Magma a
+  deriving (Eq, Show)
+
+groupByK :: Eq k => [(k, a)] -> [(k, NonEmpty a)]
+groupByK =
+    List.map (\ys -> (fst (NE.head ys), NE.map snd ys)) .
+    NE.groupBy ((==) `on` fst)
 
 prop_list :: [Int] -> Bool
 prop_list xs = (sort (nub xs) == [x | (x,()) <- toList (fromList [(x,()) | x <- xs])])
 
 prop_descList :: [Int] -> Bool
 prop_descList xs = (reverse (sort (nub xs)) == [x | (x,()) <- toDescList (fromList [(x,()) | x <- xs])])
 
+prop_fromDescList :: [(Int, A)] -> Property
```

Comment:
It's not at all clear to me what this is testing. Could you explain? Is there a simpler way to express the idea?

Reply:
This tests that given a list in decreasing order (`down_sort_xs`), the sequence we get from `fromDescList` is the same as the reverse of the sequence we would get if we group by the key and take the last key-value pair in each group.

I wouldn't mind changing the code if there is a simpler way to express this.

Reply:
Mostly need comments to explain what is being tested. Also: what's with the `()`s above? Those look mysterious.

Reply:
Added some comments.

> Also: what's with the `()`s above?

I haven't introduced those tests. It seems they want to test that `toList . fromList` sorts and nubs the keys.

Reply:
Oh, sorry, misread.

Reply:
Is it better with the comments?

Reply:
I'm going to rework the map tests to just test against `fromList*`.  
Currently the test is independent of other map functions, which will no longer be true, but it should be easier to understand what is being tested.

Reply:
Done. @treeowl let me know if you like this better.